domains/certificates/RegistryConnector/Worker/EventHandlers/MeasurementEventHandler.cs

```diff
@@ -0,0 +1,221 @@
+                productionCertificate.Id,
+                period.ToDateInterval(),
+                matchingContract.GridArea,
+                message.GSRN,
+                technology.TechCode,
+                technology.FuelCode,
+                commitment,
+                ownerPublicKey);
+
+            var transaction = issuedEvent.CreateTransaction(issuerKey);
+
+            BuildRoutingSlip(builder, transaction, commitment, matchingContract, walletDepositEndpointPosition.Value);
+
+            logger.LogInformation("Created production certificate for {Message}", message);
+        }
+        else if (matchingContract.MeteringPointType == MeteringPointType.Consumption)
+        {
+            var consumptionCertificate = new ConsumptionCertificate(
+                matchingContract.GridArea,
+                period,
+                matchingContract.MeteringPointOwner,
+                message.GSRN,
+                message.Quantity,
+                commitment.BlindingValue.ToArray());
+
+            dbContext.Add(consumptionCertificate);
+
+            var (ownerPublicKey, issuerKey) = keyGenerator.GenerateKeyInfo(message.Quantity,
+                matchingContract.WalletPublicKey, walletDepositEndpointPosition.Value, matchingContract.GridArea);
+
+            var issuedEvent = Registry.CreateIssuedEventForConsumption(
+                projectOriginOptions.RegistryName,
+                consumptionCertificate.Id,
+                period.ToDateInterval(),
+                matchingContract.GridArea,
+                message.GSRN,
+                commitment,
+                ownerPublicKey);
+
+            var transaction = issuedEvent.CreateTransaction(issuerKey);
+
+            BuildRoutingSlip(builder, transaction, commitment, matchingContract, walletDepositEndpointPosition.Value);
+
+            logger.LogInformation("Created consumption certificate for {Message}", message);
+        }
+        else
+            throw new CertificateDomainException(string.Format("Unsupported metering point type {0} for message {1}", matchingContract.MeteringPointType, message));
+
+        var routingSlip = builder.Build();
+
+        await context.Execute(routingSlip);
+        await dbContext.SaveChangesAsync(context.CancellationToken);
+    }
+
+    private static bool ShouldEventBeProduced(CertificateIssuingContract contract, EnergyMeasuredIntegrationEvent energyMeasuredIntegrationEvent)
+    {
+        if (!contract.Contains(energyMeasuredIntegrationEvent.DateFrom, energyMeasuredIntegrationEvent.DateTo))
+            return false;
+
+        if (energyMeasuredIntegrationEvent.Quantity <= 0)
+            return false;
+
+        if (energyMeasuredIntegrationEvent.Quantity > uint.MaxValue)
+            return false;
+
+        if (energyMeasuredIntegrationEvent.Quality != MeasurementQuality.Measured)
+            return false;
+
+        return true;
+    }
+
+    private void BuildRoutingSlip(RoutingSlipBuilder builder,
```

Comment:
What is RoutingSlip?

Reply:
"A routing slip specifies a sequence of processing steps called activities that are combined into a single transaction. As each activity completes, the routing slip is forwarded to the next activity in the itinerary. When all activities have completed, the routing slip is completed and the transaction is complete." - https://masstransit.io/documentation/patterns/routing-slip